waiter/src/waiter/scheduler.clj

```diff
@@ -395,50 +396,115 @@
+                                                       (str "scheduler-syncer: sync took " (- (.getMillis now) (.getMillis request-apps-time)) " ms")))]
+                        (log/trace "scheduler-syncer: querying scheduler")
+                        (if-let [service->service-instances (timers/start-stop-time!
+                                                              (metrics/waiter-timer "core" "scheduler" "app->available-tasks")
+                                                              (do-health-checks (request-available-waiter-apps scheduler)
+                                                                                (fn available [instance health-check-path]
+                                                                                  (available? instance health-check-path http-client))
+                                                                                service-id->service-description-fn))]
+                          (let [available-services (keys service->service-instances)
+                                available-service-ids (map :id available-services)]
+                            (log/debug "scheduler-syncer:" (count service->service-instances) "available services:" available-service-ids)
+                            (doseq [service available-services]
+                              (when (zero? (reduce + 0 (filter number? (vals (select-keys (:task-stats service) [:staged :running :healthy :unhealthy])))))
+                                (log/info "scheduler-syncer:" (:id service) "has no live instances!" (:task-stats service))))
+                            (loop [service-id->unhealthy-instances' {}
+                                   service-id->tracked-failed-instances' {}
+                                   instance-id->failed-health-checks' {}
+                                   scheduler-messages [[:update-available-apps {:available-apps available-service-ids :scheduler-sync-time request-apps-time}]]
+                                   [[service {:keys [active-instances failed-instances]}] & remaining] (seq service->service-instances)]
+                              (if service
+                                (let [request-instances-time (t/now)
+                                      service-id (:id service)
+                                      last-unhealthy-instances (get service-id->unhealthy-instances service-id)
+                                      service-instance-info (retrieve-instances-for-app service-id active-instances)
+                                      unhealthy-instances (:unhealthy-instances service-instance-info)
+                                      instance-id->instance-fn (fn [id instances] (first (filter #(= id (:id %)) instances)))
+                                      instances->ids-fn (fn [instances] (set (map :id instances)))
+                                      tracked-failed-instances (reduce conj (get service-id->tracked-failed-instances service-id)
```

Comment:
Track the failed instances separately from determining if there are existing failed instances.